### ISSUE TITLE
How to decode an X509 structure in an (external) provider?

### ISSUE THREAD

**Issue by baentsch** [openssl/openssl#16697]
When implementing an (OpenSSL-external) provider, it seems to be straightforward to encode into X509 (QSC/PQC) public key material using `X509_PUBKEY_set0_param`. However, I'd like to ask which API to use to _decode_ an already encoded X509 structure passed into the provider decoder as `const unsigned char **pp, long length`?

In other words, how can one create _within the provider_ an `X509_PUBKEY` from such data (without recursing into the provider decoder code)? APIs then exist to extract the key material to an OQS/EVP key structure.

A workable solution would be to call `ossl_d2i_X509_PUBKEY_INTERNAL` (as is done by all OpenSSL internal providers) but this function is not part of the public APIs... For reference, this work is part of https://github.com/open-quantum-safe/oqs-provider/issues/2.

Thanks in advance for any suggestions (maybe for an entirely different approach to decode such SPKI structure) and apologies in advance for a possible lack of understanding of the existing and available X509/ASN.1 APIs.

**Comment by t8m** [comment]
Your provider should provide a decoder implementation for the SubjectPublicKeyInfo structure of your key type. This decoder implementation will be called when an X509 ASN.1 data is being decoded. At least this is the intention how that should work.

**Comment by baentsch** [comment]
That's precisely what I did and it works perfectly when the provider calls `ossl_d2i_X509_PUBKEY_INTERNAL` to create an `X509_PUBKEY` from the data passed to the provider decoder:

```
OQSX_KEY *oqsx_d2i_PUBKEY(OQSX_KEY **a, const unsigned char **pp, long length)
{
    OQSX_KEY *key = NULL;
    X509_PUBKEY *xpk;

    // only way to re-create/decode X509 object?? TBD
    xpk = oqsx_d2i_X509_PUBKEY_INTERNAL(pp, length, NULL);

    key = oqsx_key_from_x509pubkey(xpk, NULL, NULL);
...

```

Unfortunately, that API is not public... Thus, would it be an option to change this (or at least open it to providers)? This would be to avoid re-inventing the wheel (and/or duplicating code and/or creating unintended library dependencies). 

If not desirable (?) could you point at good documentation as to how to easily decode SPKI structures (generated by `X509_PUBKEY_set0_param`)? Or would you suggest I also do "my own" SPKI _encoding_? If the latter, what APIs should be used for that? Would it really be necessary to create all this "new" code even though in principle that logic already exists within OpenSSL (cryptolib)? You can tell that a) I'm lazy :) and b) think that new code only has new bugs (that existing code hopefully doesn't).

**Comment by t8m** [comment]
I do not think we want to make ossl_d2i_X509_PUBKEY_INTERNAL() function public. At least not in 3.0.x releases since they are bugfix only.

I would just copy the ossl_d2i_X509_PUBKEY_INTERNAL implementation. It is not that big chunk of code so I do not see any real code duplication issue here.

**Comment by baentsch** [comment]
> I would just copy the ossl_d2i_X509_PUBKEY_INTERNAL implementation. 

That's what I did (to break the library dependency). Unfortunate "side effect": `ASN1_SEQUENCE` also needed to be copied in consequence and _that_ required `struct X509_pubkey_st` to be copied over, which creates a possibly dangerous dependency of this "provider" code and OpenSSL's core implementation (of `X509_PUBKEY`).

If you don't consider this "risky" too I'm fine with that solution.

**Comment by levitte** [comment]
Side note: it's a bit confusing to say X509 when you really mean X509_PUBKEY...

**Comment by levitte** [comment]
This issue is really not about the ASN.1 structure, but rather with the fact that we insist on including the decoded EVP_PKEY in it, thereby making it rather impossible to *just* decode the DER into suitable data chunks with no further interpretation and process it further.
While that was a suitable optimization, it turned out to be an unfortunate locking into the specific methodology we have before providers came along...

I would actually rather have had X509_PUBKEY be the uninterpreted contents (i.e. what we have in X509_PUBKEY_INTERNAL), and have something else for the interpretted contents, so our users (including provider authors) would have a choice.  The overall X509 structure would of course still use the interpretted structure, which I believe is possible to do.

But here we are today...  in vue of this issue, I actually think it would be useful to make X509_PUBKEY_INTERNAL publically available, possibly under another name.

**Comment by nhorman** [comment]
@levitte given the fact that this is now 3 years old, do you still think its worth exposing the proposed funtion, or shall we close this issue?

**Comment by t8m** [comment]
Closing. If there is a concrete and actionable feature request, please open it as a new issue to avoid confusion.
